src/pr-allowed.js

```diff
@@ -0,0 +1,109 @@
+let yaml = require('js-yaml');
+
+const DEFAULT_POLICY = 'collaborators';
+
+async function prAllowed(options) {
+  switch (await getRepoPolicy(options)) {
+    case 'collaborators':
+      return await isCollaborator(options);
+
+    case 'public':
+      return true;
+
+    default:
+      return false;
+  }
+}
+
+/**
+ * Get the repo's "policy" on pull requests, by fetching .taskcluster.yml from the default
+ * branch, parsing it, and looking at its `allowPullRequests`.
+ */
+async function getRepoPolicy({login, organization, repository, instGithub, debug}) {
+  // first, get the repository's default branch
+  let repoInfo = await instGithub.repos.get({owner: organization, repo: repository});
+  let branch = repoInfo.default_branch;
+
+  // load .taskcluster.yml from that branch
+  let taskclusterYml;
+  try {
+    let content = await instGithub.repos.getContent({
+      owner: organization,
+      repo: repository,
+      path: '.taskcluster.yml',
+      ref: branch,
+    });
+    taskclusterYml = yaml.safeLoad(new Buffer(content.content, 'base64').toString());
+  } catch (e) {
+    if (e.code === 404) {
+      return DEFAULT_POLICY;
+    }
+    throw e;
+  }
+
+  // consult its `allowPullRequests` field
+  return taskclusterYml['allowPullRequests'] || DEFAULT_POLICY;
+}
+
+async function isCollaborator({login, organization, repository, sha, instGithub, debug}) {
+  if (login === organization) {
+    debug(`Checking collaborator: ${login} === ${organization}: True!`);
+    return true;
+  }
+
+  // If the user is in the org, we consider them
+  // qualified to trigger any job in that org.
+  try {
+    await instGithub.orgs.checkMembership({
+      org: organization,
+      owner: login,
+    });
+    debug(`Checking collaborator: ${login} is a member of ${organization}: True!`);
+    return true;
+  } catch (e) {
+    if (e.code == 404) {
+      // Only a 404 error means the user isn't a member
+      // anything else should just throw like normal
+    } else {
+      throw e;
+    }
+  }
+
+  // GithubAPI's collaborator check returns an error if a user isn't
+  // listed as a collaborator.
+  try {
+    await instGithub.repos.checkCollaborator({
+      owner: organization,
+      repo: repository,
+      collabuser: login,
+    });
+    // No error, the user is a collaborator
+    debug(`Checking collaborator: ${login} is a collaborator on ${organization}/${repository}: True!`);
+    return true;
+  } catch (e) {
+    if (e.code == 404) {
+      // Only a 404 error means the user isn't a collaborator
+      // anything else should just throw like normal
+    } else if (e.code == 403) {
+      let msg = `Taskcluster does not have permission to check for repository collaborators.
+        Ensure that it is a member of a team with __write__ access to this repository!`;
```

Comment:
Really, I think this whole `else if` doesn't really make sense anymore in a post-taskclusterrobot world. We can probably get rid of it.